commons/src/main/java/fr/free/nrw/commons/category/CategorizationFragment.java

```diff
@@ -87,113 +95,72 @@ public void writeToParcel(Parcel parcel, int flags) {
-                        Log.d(TAG, "GPS cats found in CategorizationFragment.java" + MwVolleyApi.getGpsCat().toString());
-                        List<String> gpsItems = new ArrayList<String>(MwVolleyApi.getGpsCat());
-                        Log.d(TAG, "GPS items: " + gpsItems.toString());
 
-                        mergedItems.addAll(gpsItems);
-                    }
+    protected void setCatsAfterAsync(ArrayList<String> categories, String filter) {
 
-                    mergedItems.addAll(items);
-                }
-                catch (RemoteException e) {
-                    // faaaail
-                    throw new RuntimeException(e);
-                }
-                Log.d(TAG, "Merged items: " + mergedItems.toString());
-                return mergedItems;
-            }
-            
-            if(categoriesCache.containsKey(filter)) {
-                return categoriesCache.get(filter);
-            }
-            MWApi api = CommonsApplication.createMWApi();
-            ApiResult result;
-            ArrayList<String> categories = new ArrayList<String>();
-            try {
-                result = api.action("query")
-                        .param("list", "allcategories")
-                        .param("acprefix", filter)
-                        .param("aclimit", SEARCH_CATS_LIMIT)
-                        .get();
-            } catch (IOException e) {
-                throw new RuntimeException(e);
+        ArrayList<CategoryItem> items = new ArrayList<CategoryItem>();
+        HashSet<String> existingKeys = new HashSet<String>();
+        for(CategoryItem item : categoriesAdapter.getItems()) {
+            if(item.selected) {
+                items.add(item);
+                existingKeys.add(item.name);
             }
-
-            ArrayList<ApiResult> categoryNodes = result.getNodes("/api/query/allcategories/c");
-            for(ApiResult categoryNode: categoryNodes) {
-                categories.add(categoryNode.getDocument().getTextContent());
+        }
+        for(String category : categories) {
+            if(!existingKeys.contains(category)) {
+                items.add(new CategoryItem(category, false));
             }
+        }
 
-            categoriesCache.put(filter, categories);
+        categoriesAdapter.setItems(items);
+        categoriesAdapter.notifyDataSetInvalidated();
+        categoriesSearchInProgress.setVisibility(View.GONE);
 
-            return categories;
+        if (categories.size() == 0) {
+            if(TextUtils.isEmpty(filter)) {
+                // If we found no recent cats, show the skip message!
+                categoriesSkip.setVisibility(View.VISIBLE);
+            } else {
+                categoriesNotFoundView.setText(getString(R.string.categories_not_found, filter));
+                categoriesNotFoundView.setVisibility(View.VISIBLE);
+            }
+        } else {
+            categoriesList.smoothScrollToPosition(existingKeys.size());
```

Comment:
Hm, sorry, not quite clear on this... do you mean that you think we should set catNotFoundView and catSkip to 'not visible' when cat.size() != 0?